src/app/app.component.ts

```diff
@@ -0,0 +1,57 @@
+import { Component, OnInit } from '@angular/core';
+import { CoffeeShopsService } from './services/coffeeShops.service';
+import {  retry, switchMap } from 'rxjs';
+import { CoffeeShopModel } from './models/coffeeShop.model';
+import { FormControl, FormGroup, Validators } from '@angular/forms';
+
+@Component({
+  selector: 'app-root',
+  templateUrl: './app.component.html',
+  styleUrls: ['./app.component.scss']
+})
+export class AppComponent implements OnInit {
+  coffeeShops = new Array<CoffeeShopModel>();
+  coordinatesFormControl = new FormGroup({
+    xCoordinate: new FormControl([], Validators.required),
+    yCoordinate: new FormControl([], Validators.required)
+  });
+
+  constructor(private coffeeShop: CoffeeShopsService) {
+  }
+
+  ngOnInit(): void {
+    this.getCoffeeShops();
+  }
+
+  getCoffeeShops(): void {
+    this.coffeeShop.getToken().pipe(
+      switchMap((result: any) => {
+        return this.coffeeShop.getCoffeeShops(result.token);
+      }), retry(5)).subscribe(result => this.coffeeShops = result);
+
+  }
+
+  getNearbyCS(): void {
+    for (let coffeeShop of this.coffeeShops) {
+      coffeeShop.distanceFromUser = Number(this.distance(this.coordinatesFormControl.value.xCoordinate, this.coordinatesFormControl.value.yCoordinate,
+        Number(coffeeShop.x), Number(coffeeShop.y)));
+    }
+    this.coffeeShops.sort((a: CoffeeShopModel, b: CoffeeShopModel): number => a.distanceFromUser - b.distanceFromUser);
+    this.coffeeShops = this.coffeeShops.slice(0, 3);
```

Comment:
I like the concise way of sorting this by just using the difference between the two distances :+1:. But this `sorting`, and the `slice` appear to be mutating the `this.coffeeShops` value you get from the API, what if other functions depended on the original coffeeShops array? Or what if I wanted to see the initial 5 coffee shops, now it looks like I am stuck on 3 sorted ones until I refresh the page.

Reply:
Yes, I know, I wanted to implement an ngrx/store, to store the initial values, and put in a basic refresh button, or just give the possiblity for the user to pass new values for X and Y. In this case, it's counter productive.